### ISSUE TITLE
AIC Editor: Limit a specific kind of defense units

### ISSUE THREAD

**Issue by Halbarad1** [Sh0wdown/UnofficialCrusaderPatch#400]
Hi,
I improved the original Sheriff by among other things giving him more defense units in his AIC file. 
I wanted him to build more Crossbowmen and European Archers. 
Problem: He builds way to much fire throwers, especially when starting with 40 000.
Suggestion: **Would be nice if there could be a parameter to limit specific defense units numbers and not only total defense units number.**

**Comment by Starshine575** [comment]
Yeah thats true I agree with that. At the moment I try to change and improve all the AIs and also change there castles. I wanted to give the snake one ot two firethrower, but the only way to achive this is by adding one to the attacking units. 🤔
But this isn't really a good solution. 😕

**Comment by Truetobi91** [comment]
Usually I'd change the combination of defensive troops the AI builds e.g. adding more of the same type to the defense unit slots to increase the ratio the AI builds of these type. 

However Sheriff already has a decent combination of 3x Crossbow and 1x Fire Thrower. The problem you describe reminds me of a different problem Sheriff is known for. He recruits way to many crossbowman as raid units so he has too few left for defense if he even starts stacking them up. 

You might want to take a look here where I described the recruit probabilities and especially the Sheriff problem. #332

**Comment by Sh0wdown** [comment]
Adding unit specific limits is kinda complex since they should be adjustable via .aic files I guess?
Otherwise you can edit the unit combination like @Truetobi91 suggested, e.g. 7x Crossbow and 1x Fire Thrower.

**Comment by Halbarad1** [comment]
@Sh0wdown yes it should be adjustable via .aic files
The suggestion of  @Truetobi91 doesn't work that good, my AIC File right now has this kind of defense troops for sheriff:
Crossbowman, Crossbowman, EuropArcher, EuropArcher, Crossbowman, Crossbowman, FireThrower, Assassin.
After half an hour against other AIs he had about 29 Firethrowers and about 12 Crossbowmen, although nearly all of the Firethrowers were placed together with a Crossbowmen on a tower. Giving him one time arabien archers instead of Crossbowmen does lower the amount of Fire Throwers by far.
I tried to fix this problem by letting him build mercenary post as his last building, but it does only help in the beginning when playing with much gold. When playing with low amount of start gold, he builds round around 48 European bowmen and 12 Crossbowmen (and probably also about 12 Fire Throwers).

So the amount of defense troops seems to depend more on the amount of gold, weapons, ... the AI has at the moment and the building costs of the units than on the  aic file right now.

**Comment by Krarilotus** [comment]
I wonder, if it makes a difference if you let another AI play with the same AIC Array and Castles. Try to put your sherrif into the wolf/saladin/caliph/abbot/rat/frederick maybe?
I might be wrong but i woud think, that maybe some kind of behaviour is still hardcoded into their Characteristics. Maybe some of the Unknown Values define a ratio for arabian to european units? We coud try to set unknown values to obvious ratios and see if something changes.

**Comment by Halbarad1** [comment]
@Krarilotus interesting idea but it didn't seemed to change something. The first time the wolf-sheriff really had more crossbowmen than the sheriff but in every other game it didn't seem to be true. Also, Wolf AI starts with more Start-Crossbowmen at the beginning than Sheriff AI (at 10000 Wolf has 8 and Sheriff had 3 or something like this). If someone wants to try him/ it out you can do that of course. Its a full working AI and would now earn the name "King Aerys" more than "Sheriff of Nottingham", but it is not as perfect as I want it to be. 

[Sheriff by )-(albarad.zip](https://github.com/Sh0wdown/UnofficialCrusaderPatch/files/3343095/Sheriff.by.-.albarad.zip)
![Sheriff3](https://user-images.githubusercontent.com/44904062/60399544-5108a000-9b66-11e9-8e3a-1d0f66797239.png)
![Sheriff](https://user-images.githubusercontent.com/44904062/60399545-5108a000-9b66-11e9-952d-db313f5072fc.png)
![Sheriff2](https://user-images.githubusercontent.com/44904062/60399546-5108a000-9b66-11e9-9887-bcec947e3ef6.png)

**Comment by Krarilotus** [comment]
Have you tried putting in only 4 firethrower spots, 10 crossbowman spots and maybe about 4 european archer spots?
I also think, that him having that many firethrowers might be duo to the fact, that they dont get focussed as much. So he recruits the defense troops by the given probabilities of 3 crossbowman/eu arch/firethrower but the archers and firethrowers might be on more defensive positions/dont get focussed as much, so they survive longer and stack more?
But i could be wrong ;P
